src/main/java/org/junit/Assert.java

```diff
@@ -955,4 +955,46 @@ public static void assertEquals(Object[] expecteds, Object[] actuals) {
             Matcher<? super T> matcher) {
         MatcherAssert.assertThat(reason, actual, matcher);
     }
+
+    /**
+     * This interface facilitates the use of expectThrows from Java 8. It allows
+     * method references to void methods (that declare checked exceptions) to be
+     * passed directly into expectThrows without wrapping. It is not meant to be
+     * implemented directly.
+     */
+    public interface ThrowingRunnable {
+        void run() throws Throwable;
+    }
+
+    /**
+     * Asserts that {@code runnable} throws an exception when executed. If it
+     * does, the exception object is returned. If it does not, an
+     * {@link AssertionError} is thrown.
+     *
+     * @param runnable A function that is expected to throw an exception when executed
+     * @return The exception thrown by {@code runnable}
+     */
+    public static Throwable expectThrows(ThrowingRunnable runnable) {
```

Comment:
I strongly feel this should take a class object of the expected type, for all the reasons that I gave as #/1153

While Arrange-Act-Assert is a good framework to keep in mind when writing tests, it isn't the only lens we should use when we design APIs.

A common thing I think about when designing an API is "Make Interfaces Easy to Use Correctly and Hard to Use Incorrectly" (see http://programmer.97things.oreilly.com/wiki/index.php/Make_Interfaces_Easy_to_Use_Correctly_and_Hard_to_Use_Incorrectly).

The API here encourages code that provides less than useful error messages when the code under test doesn't meet the tester's expectations. For example:

    assertTrue(expectThrow(client::invoke) instanceof RuntimeException); // horrible message
    IllegalArgumentException e
        = (IllegalArgumentException) expectThrow(client::invoke); // not much better

Worse, the test could pass for the wrong reason

    expectThrow(client::invoke); // what the method starts throwing StackOverflowError ?

It also makes it difficult to do simple things:

    assertEquals(3, ((MyException) expectThrow(client::invoke)).getErrorCode());

Both of these are solved if we pass in the expected type:

    ClientException e = expectThrow(ClientException.class, client::invoke);
    assertThat(e.getMessage()).contains("connection is closed");

    assertEquals(3, expectThrow(MyException.class, client::invoke).getErrorCode());